libp2p/NodeTable.cpp

```diff
@@ -401,49 +418,60 @@ void NodeTable::onPacketReceived(
         {
             case Pong::type:
             {
-                auto in = dynamic_cast<Pong const&>(*packet);
-                // whenever a pong is received, check if it's in m_evictions
-                bool found = false;
-                NodeID leastSeenID;
-                EvictionTimeout evictionEntry;
-                DEV_GUARDED(x_evictions)
-                { 
-                    auto e = m_evictions.find(in.sourceid);
-                    if (e != m_evictions.end())
-                    {
-                        if (e->second.evictedTimePoint + c_reqTimeout >=
-                            std::chrono::steady_clock::now())
-                        {
-                            found = true;
-                            leastSeenID = e->first;
-                            evictionEntry = e->second;
-                            m_evictions.erase(e);
-                        }
-                    }
+                auto pong = dynamic_cast<Pong const&>(*packet);
+                auto const sourceId = pong.sourceid;
+
+                // validate pong
+                auto const sentPing = m_sentPings.find(sourceId);
+                if (sentPing == m_sentPings.end())
+                {
+                    LOG(m_logger) << "Unsolicited PONG from " << _from.address().to_string() << ":"
+                                  << _from.port();
+                    return;
                 }
 
-                if (found)
+                if (pong.echo != sentPing->second.second)
                 {
-                    if (auto n = nodeEntry(evictionEntry.newNodeID))
-                        dropNode(n);
-                    if (auto n = nodeEntry(leastSeenID))
-                        n->pending = false;
+                    LOG(m_logger) << "Invalid PONG from " << _from.address().to_string() << ":"
+                                  << _from.port();
+                    return;
                 }
-                else
+
+                m_sentPings.erase(sentPing);
+                auto const sourceNodeEntry = nodeEntry(sourceId);
+                assert(sourceNodeEntry.get());
```

Comment:
This assert fails.

Reply:
It shouldn't, because we checked that `m_sendPings` had this `sourceId` and we always add a node to `m_sentPings` after it is already in `m_allNodes`

Reply:
> It shouldn't, because we checked that `m_sendPings` had this `sourceId` and we always add a node to `m_sentPings` after it is already in `m_allNodes`

Well...
```
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##d5f4476e…@149.28.150.71:30303
DEBUG 12-10 18:15:27 p2p  discov PONG from ##d5f4476e… 149.28.150.71:30303
DEBUG 12-10 18:15:27 p2p  discov Noting active node: ##d5f4476e… 149.28.150.71:30303
DEBUG 12-10 18:15:27 p2p  discov Sending Ping to ##8152c411…@35.188.130.212 UDP 30303 TCP 30303
DEBUG 12-10 18:15:27 p2p  discov Received Ping from ##d5f4476e…@149.28.150.71:30303
DEBUG 12-10 18:15:27 p2p  discov Sending Pong to ##d5f4476e…@149.28.150.71:30303
DEBUG 12-10 18:15:27 p2p  discov Noting active node: ##d5f4476e… 149.28.150.71:30303
DEBUG 12-10 18:15:27 p2p  discov Sending Ping to ##8152c411…@35.188.130.212 UDP 30303 TCP 30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##d9e5dc65…@150.109.58.135:30303
DEBUG 12-10 18:15:27 p2p  discov PONG from ##d9e5dc65… 150.109.58.135:30303
DEBUG 12-10 18:15:27 p2p  discov Noting active node: ##d9e5dc65… 150.109.58.135:30303
DEBUG 12-10 18:15:27 p2p  discov Sending Ping to ##8152c411…@35.188.130.212 UDP 30303 TCP 30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##67f8e81c…@52.83.138.158:30307
DEBUG 12-10 18:15:27 p2p  discov PONG from ##67f8e81c… 52.83.138.158:30307
DEBUG 12-10 18:15:27 p2p  discov Noting active node: ##67f8e81c… 52.83.138.158:30307
DEBUG 12-10 18:15:27 p2p  discov Sending Ping to ##8152c411…@35.188.130.212 UDP 30303 TCP 30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov p2p.nodes.drop ##8152c411…
DEBUG 12-10 18:15:27 p2p  discov Noting active node: ##b57b2647… 46.101.253.211:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov Invalid PONG from 35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  net    p2p.host.nodeTable.events.NodeEntryDropped ##8152c411…
DEBUG 12-10 18:15:27 p2p  net    p2p.host.nodeTable.events.nodeEntryAdded ##b57b2647…
DEBUG 12-10 18:15:27 p2p  net    p2p.host.peers.events.peerAdded ##b57b2647… 46.101.253.211 UDP 30303 TCP 30303
DEBUG 12-10 18:15:27 p2p  discov Received Pong from ##8152c411…@35.188.130.212:30303
DEBUG 12-10 18:15:27 p2p  discov ASSERT FAILURE 35.188.130.212:30303
aleth-bootnode: /home/chfast/Projects/ethereum/aleth/libp2p/NodeTable.cpp:446: virtual void dev::p2p::NodeTable::onPacketReceived(dev::p2p::UDPSocketFace*, const endpoint&, dev::bytesConstRef): Assertion `sourceNodeEntry.get()' failed.
```

Reply:
That's interesting, the same node sends us several invalid PONGs, and then the valid one...

Reply:
So I think what happens is the node is being evicted via eviction mechanism and then the problem is its record in `m_sentPings` is not deleted at the same time.

Now I realize that eviction and timeout handling is basically the same thing. Both drop a node after PING timeouts, the only difference that in eviction the new node is added after dropping the old one.
Maybe I'll try to simplify it by unifying them...

Reply:
Good idea.

Reply:
We have to wait for this fix, right?

Reply:
Yes, this currently fails for me, too

Reply:
> So I think what happens is the node is being evicted via eviction mechanism and then the problem is its record in `m_sentPings` is not deleted at the same time.
> 
> Now I realize that eviction and timeout handling is basically the same thing. Both drop a node after PING timeouts, the only difference that in eviction the new node is added after dropping the old one.
> Maybe I'll try to simplify it by unifying them...

This is exactly the case. Its based on kademlia paper and algorithm and there only needs to be one discovery process and separate timers/expirations aren't necessary.

Reply:
Right. Yet there's still one old separate timeout mechanism for `FindNode` requests. Maybe we'll simplify it later...

Reply:
@chfast Please check this with the latest changes, should be fixed.

Reply:
I did some commits ago. It's fixed.